added some extra hacky code that causes it to work with both my strip…

gerth2 opened the pull request:
I have the following device: 
https://www.amazon.com/gp/product/B01DY56N8U/ref=oh_aui_detailpage_o01_s00?ie=UTF8&psc=1

It shows up as having "V1" firmware in the Magic Home app. 

A packet capture shows an additional byte is sent during color and rgb updates:

    0x31 0x00 0x00 0x00 W 0x00 0x0F 0x0F CSUM
    or
    0x31 R G B 0x00 0x00 0xF0 0x0F CSUM

At present, I am triggering adding the (apparently) extra byte for the V1 RGBW is triggered by the same conditions as the "BadRGBW", which happens to be what the problematic device above registers as. I am unsure if this is the only device which gets the "BadRGBW" classification. I also did not immedeately see any indication of determining the V1 vs. V3 firmware version in the packets exchanged. 

I've got one commit in my forked repo where I have the color setting and white intensity setting working for both my strips (the other is https://www.amazon.com/gp/product/B01JZ2SI6Q/ref=oh_aui_detailpage_o02_s00?ie=UTF8&psc=1). 

Since I'm somewhat new around here... What would be the desired path forward for integrating such changes?

gerth2 commented:
Note this appears related to https://github.com/beville/flux_led/issues/5 from the parent repo.

claudobahn commented:
@Danielhiversen I have a 5 channel (RGB, CW, WW) capable controller and the change in this PR allowed me to get things working. Without this, I was only able to turn on/off.

@gerth2 are you sure your extra byte belongs ahead of the special byte?

gerth2 commented:
I think so? What I need to do is swap to the full five channel mode and see
what the packet captures look like. I suspect the extra byte is related to
that. I will hopefully have some time this weekend to give it a shot.

On May 19, 2017 3:59 PM, "claudobahn" <notifications@github.com> wrote:

> @Danielhiversen <https://github.com/danielhiversen> I have a 5 channel
> (RGB, CW, WW) capable controller and the change in this PR allowed me to
> get things working. Without this, I was only able to turn on/off.
>
> @gerth2 <https://github.com/gerth2> are you sure your extra byte belongs
> ahead of the special byte?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Danielhiversen/flux_led/pull/42#issuecomment-302810061>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEXw7vUMjOY0WI5aGCoJQDksVfuu92tEks5r7gLFgaJpZM4NXfme>
> .
>

claudobahn commented:
Actually, I think the extra byte comes after what's currently called the special byte.

A packet capture of setting the RGB value from the Magic Color app:

> 31 bc c1 ff 00 00 f0 0f ac
> 31 - command
> bc - red
> c1 - green
> ff - blue
> 00 - warm white (unused, because on RGB screen)
> 00 - cold white (unused, because on RGB screen)
> f0 - set RGB only (white values are ignored)
> 0f - terminator?
> ac - checksum

A packet capture of setting the white values (warm/cold) from the Magic Color app:

> 31 00 00 00 d9 8a 0f 0f b2
> 31 - command
> 0 - red
> 0 - green
> 0 - blue
> d9 - warm white
> 8a - cold white
> 0f - set whites only
> 0f - terminator?
> b2 - checksum

This is just a guess, but presumably it's also possibly to use ff instead of f0 and 0f, and provide all 5 channels in the same message??

claudobahn commented:
btw, what does the packet your controller sends back look like? Mine's like this:
> 81 25 24 61 21 06 7d ff f9 f2 01 00 f0 aa
> 81 25 23 61 21 06 ff 18 1c f2 01 00 f0 67
> 81 25 23 61 21 06 38 05 06 f2 01 00 f0 77
> 81 25 23 61 21 06 38 05 06 f9 01 00 0f 9d

From what I can tell (using last one):
81,25 some kind of identifier
23 (or 24) off/on (as usual)
61 effect
21 unknown
06 speed
38 red
05 green
06 blue
f9 warm white
01 unknown, doesn't seem to change?
00 cold white
0f (white mode, when it was f0, I had last set a color?)
9d checksum

gerth2 commented:
So here's an example of what I get for state (returned):

0x81 0x25 0x23 0x61 0x21 0x10 0xa0 0x96 0x00 0x02 0x01 0x00 0xf0 0x84

I presume 0x81/0x25 is just a special prefix to identify the packet's purpose.

3rd packet also toggles between 0x23 (on) and 0x24 (off)

I haven't tried fiddling with the effect/speed yet, todo.

My RGB, CW, and WW bytes do line up with yours. Also, the white/color mode byte (second to last) behavior aligns with what you described.

One side note on this is that I think green and red were flipped in label on the box - even the MagicHome app gets them wrong. I just re-did the wiring to make it line up with the packets sent both by this python code, and the MagicHome app. Ideally I think there would be a way to provide the remapping either in this class or in HomeAssistant somehow - editing a config .yaml file is generally easier than breaking out the soldering equipment.

gerth2 commented:
I was able to replicate your sent packet results as well.

Sent Packet protocol (what I"m seeing at least):

    Header (0x31 Always)
    RED
    GRN
    BLU
    WW
    CW
    Mode (0x0F for whites, 0xF0 for colors)
    Unknown (0x0F Always)
    Checksum

So, my "Extra byte" is actually the Cool-white channel. Next question: Is there any way to differentiate 4 and 5 channel controllers from either the status packet or the initial connection messages?

gerth2 commented:
@claudobahn , per your earlier comment of "it's also possibly to use ff " - you're close - my experiments are showing that using 0x00 for this byte allows all five channels to be set at once. Apparently that byte means something like "Ignore commands from color/noncolor channels" ?

So, what this means is our controller isn't technically a "badRGBW" one, even though the second byte in the status message is 0x25.

In the interest of not breaking other users, I've updated the state parsing logic to have two levels: If we see the specific 0x8125 header AND a status length of exactly 14, the script will assume we're working with our 5-channel v1 controller. Otherwise, the script uses the old parsing logic to decide module capabilities.  I wouldn't be opposed to doing it another way, but without knowing full details on the operation of other devices, I think this is a good approach.

I've added a few more things that were helping me debug:
  - Command line cool-white alone control
  - Updated -c option to allow for up to five colors numbers specified (last color commands cool white channel if available)
  - Added command line switch to enable/disable verbose print messages (which I found useful while debugging protocols)
  -

Danielhiversen commented on flux_led/__main__.py:
```diff
@@ -497,7 +497,7 @@ def __str__(self):
 
 class WifiLedBulb():
-    def __init__(self, ipaddr, port=5577, timeout=5):
+    def __init__(self, ipaddr, verbose=False, port=5577, timeout=5 ):
```
Add verbose as last argument

Danielhiversen commented on flux_led/__main__.py:
```diff
@@ -516,9 +517,12 @@ def __init__(self, ipaddr, port=5577, timeout=5):
 
         self.connect(2)
         self.update_state()
+        
```
Remove blank lines

Danielhiversen commented on flux_led/__main__.py:
```diff
@@ -1705,11 +1803,11 @@ def main():
             print("Unable to connect to bulb at [{}]: {}".format(info['ipaddr'],e))
             continue
-
+            
```
Remove spaces

Danielhiversen commented on flux_led/__main__.py:
```diff
@@ -1775,7 +1876,7 @@ def main():
                 print("  Timer #{}: {}".format(num, t))
             print("")
-
+        
```
Remove spaces

SmaginPV commented:
I have the same type of controller, as @gerth2 have.
In MagicHome app I choose RGBWW device type to be able to control my led. Unfortunately this type of controller is not supported by HASS yet. I can only turn on/off.

tchellomello commented:
Adding on my plate to test it.

Danielhiversen commented:
Will merge #46 instead
